File: wrdk/loader/loader.c
/** A simple serial based boot loader.

    @author Michael Hope
*/
#include "loader.h"

/* Protocol specification:

   The protocol is a synchronous command/response system where the
   loader receives a command from the client, executes it, then returns
   the response.

   Commands and responses are sent in frames.  A frame starts with the
   first character received and ends with a \n (0x0A).  \r (0x0D) is
   ignored.

   Sending Ctrl-C at any time will interrupt the loader, cause any
   current command to be discarded, and generate a '!' response.

   Most frames are printable ASCII.  Some frames such as memory store
   have a binary variant.

   Special characters are escaped by sending the escape character '}'
   (0x7D) then the actual character XORed with 0x20.  \n, \r, Ctrl-C,
   and } must be escaped.  Note that this is only needed on binary
   frames.
   
   Commands are made up of a command ID and then optional space
   separated arguments.  Numbers are in hex with lower case for the
   alpha values.

   'Gets' use lower case command IDs.  'Sets' use upper case. See
   dispatch() for the list of commands.

   Responses may be either the data, 'k' for success, 'e' for error,
   or '?' for unrecognised command.
*/

/* Design notes:

   There's no checksum as a serial link is actually very reliable.
   Having no checksum means you can also talk to the loader manually
   over a terminal.

   There's no value or bounds checking.  This keeps the loader small
   and, as you're already using a system with direct memory and
   program access, there's no extra security risk.

   'uint' is used instead of more specific types.  One architecture
   the author is interested in doesn't have good casting or signed
   comparisons instructions.
*/

/** Stringify a macro */
#define STR(_x) #_x

/** All of the hex characters.  Used in encoding */
static const uint8_t nibbles[] = "0123456789abcdef";

/** Shared state across the module */
struct
{
    /** Current position in the input stream */
    uint8_t* pat;
    /** End of the command.  Used for binary packets. */
    uint8_t* pend;
} q;

/** The input buffer.  Done outside of q to allow putting it in a
    different segment */
uint8_t command[LOADER_INPUT_SIZE];

/** Decode a hex nibble */
static uint denibble(uint ch)
{
    /* Note that there's no range checking */
    if (ch >= 'a')
    {
        return ch - 'a' + 10;
    }
    else if (ch >= 'A')
    {
        return ch - 'A' + 10;
    }
    else
    {
        return ch - '0';
    }
}

/** Writes out the given character */
void loader_putch(uint ch)
{
    port_putch(ch);
}

/** Writes out the given null terminated string */
void loader_puts(const char* pmsg)
{
    for (; *pmsg != '\0'; pmsg++)
    {
        loader_putch(*pmsg);
    }
}

/** Writes out the given value as a hex byte */
void loader_putb(uint byte)
{
    loader_putch(nibbles[(byte >> 4) & 0x0F]);
    loader_putch(nibbles[(byte >> 0) & 0x0F]);
}

/** Start the response */
static void start()
{
    /* Any future extensions such as framing or checksums go here */
}

/** Finish up the response */
static void end()
{
    /* Write out the end of message */
    port_putch('\n');
}

/** Respond with OK */
static void ok()
{
    loader_putch('k');
}

/** Get the next hex value from the command */
uint loader_get()
{
    uint v = 0;

    while (1)
    {
        uint got = *q.pat++;

        /* Values are separated by space or end of string */
        if (got == ' ' || got == '\0')
        {
            return v;
        }

        v <<= 4;
        v |= denibble(got);
    }
}

static uint receive()
{
    uint8_t* pat = command;
    uint escaped = 0;

    /* Read in the whole of the packet */
    while (1)
    {
        uint ch = port_getch();

        if (ch == '\003')
        {
            /* Ctrl-C.  Discard everything */
            return ch;
        }
        else if (ch == '\r')
        {
            q.pend = pat;
            q.pat = command + 1;

            /* Null terminate to make things easier */
            *pat = '\0';

            /* Note that a zero length command has \0 as the first
               byte */
            return command[0];
        }
        else if (ch == '\n')
        {
            /* Eat so that \r\n from a terminal is fine */
        }
        else if (ch == '}')
        {
            /* PPP style escaping.  Most of the commands are plain
               text but there's no harm in de-escaping all types of
               command.
            */
            escaped = 0x20;
        }
        else
        {
            *pat++ = ch ^ escaped;
            escaped = 0;

            /* Note that there's no buffer overflow checking */
        }
    }
}

void dispatch(uint id)
{
    switch (id)
    {
    case '\003':
        /* Ctrl-C.  Discard the frame and re-sync
           Response: '!'
        */
        loader_putch('!');
        break;

    case 'm':
    {
        /* Read memory.
           Arguments: start address, count in bytes
           Response: bytes in hex
        */
        uint8_t* p = (uint8_t*)loader_get();
        uint8_t* pend = p + loader_get();

        for (; p != pend; p++)
        {
            loader_putb(*p);
        }

        break;
    }
    case 'M':
    {
        /* Write memory
           Arguments: start address, data as hex bytes
           Response: k
        */
        uint8_t* p = (uint8_t*)loader_get();
        uint8_t* pend = p + loader_get();

        for (; p != pend; p++)
        {
            *p = (denibble(*q.pat++) << 4) | denibble(*q.pat++);
        }

        ok();
        break;
    }
    case 'X':
    {
        /* Write memory in binary mode
           Arguments: start address, rest as binary data bytes
           Response: k
        */
        uint8_t* p = (uint8_t*)loader_get();
        const uint8_t* pat;

        for (pat = q.pat; pat != q.pend; pat++)
        {
            *p++ = *pat;
        }

        ok();
        break;
    }
    case 'u':
    {
        /* Checksum a block of memory
           Arguments: start address, count in bytes
           Response: 16 bit modulo 65536 checksum
        */
        uint8_t* p = (uint8_t*)loader_get();
        uint8_t* pend = p + loader_get();
        uint sum = 0;

        for (; p != pend; p++)
        {
            sum += *p;
        }
        loader_putb((sum >> 8) & 0xFF);
        loader_putb((sum >> 0) & 0xFF);
        break;
    }
    case 'c':
    {
        /* Change execution
           Argument: address to start executing at
           Response: k
        */
        void (*pstart)(void) = (void (*)(void))loader_get();
        (*pstart)();
        ok();
        break;
    }
    case 'B':
    {
        /* Set the baud rate in a device specific way
           Argument: New value for the divisor
           Response: k
        */
        port_set_baud(loader_get());
        ok();
        break;
    }
    case '?':
        /* Get capabilities
           Response: Space separated capabilities.  First character of
           each is the capability type.

           v: protocol version.  Always the first field
           t: host CPU type
           c: command capacity in decimal
        */
        loader_puts("v1 t" STR(LOADER_TYPE) " c" STR(LOADER_INPUT_SIZE));
        break;

    default:
        loader_putch('?');
        break;
    }
}

/** Main entry point */
void loader()
{
    for (;;)
    {
        uint code = receive();

        start();
        dispatch(code);
        end();

        port_progress();
    }
}

File: wrdk/loader/loader.py
"""Python module that implements the loader protocol over a serial port.
"""
import serial
import time
import sys

class LoaderException(Exception):
    pass

class Loader:
    """Host side implementation of the serial loader protocol.  Use to
    load programs into the device and start execution.
    """
    def __init__(self, port, rate=115200):
        self._s = serial.Serial(port, rate, timeout=0.010)
        
    def encode(self, msg):
        """Encode a message"""
        return '%s\r' % msg

    def send(self, msg):
        """Encode and send a message"""
        enc = self.encode(msg)
        self._s.write(enc)

    def get_nibble(self, ch):
        """Convert a nibble to an integer"""
        return '0123456789abcdef'.index(ch)

    def get(self):
        """Get and return the response, throwing on no response."""
        msg = ''
        end = time.time() + 1

        while True:
            got = self._s.read(4096)

            for ch in got:
                if ch == '\n':
                    pass
                elif ch == '\r':
                    return msg
                else:
                    msg += ch
        
            if time.time() > end:
                raise LoaderException('No response')

    def check_ok(self):
        """Check that the loader responded with OK"""
        got = self.get()

        if got != 'k':
            raise LoaderException('Target returned "%s"' % got)

    def read(self, address, count):
        """Read count bytes starting at address and return as a string"""
        self.send('m%x %x' % (address, count))
        return self.get()

    def write(self, address, data):
        """Write the given data to the given address"""
        payload = ''.join('%02x' % ord(ch) for ch in data)
        self.send('M%x %x %s' % (address, len(data), payload))
        self.check_ok()

    def writebin(self, address, data):
        """Write the given data to the given address in binary transfer mode"""
        special = '\r\n\003}'
        payload = ''.join('}%s' % chr(ord(x) ^ 0x20) if x in special else x for x in data)
        self.send('X%x %s' % (address, payload))
        self.check_ok()

    def checksum(self, address, count):
        """Get the sixteen bit checksum of count bytes starting at the
        given address
        """
        self.send('u%x %x' % (address, count))
        return int(self.get(), 16)

    def go(self, address):
        """Start execution at the given address"""
        self.send('c%x' % address)

        try:
            self.check_ok()
        except:
            # Ignore the exception after starting the program
            pass

    def set_baud(self, bps):
        """Change the baud rate"""
        fbrclk = 48000000
        div = 1
        divmd = 8
        brr = (((((10 * fbrclk) / div / divmd)/(2 * bps)) + 5) / 10 - 1)
        self.send('B%x' % brr)

    def load(self, address, bin):
        """Load the data at the given starting address.  Chunks the
        data up and verifies the checksum.

        Throws on any error.
        """
        chunk = 256

        for i in range(0, len(bin), chunk):
            self.writebin(address + i, bin[i:i+chunk])
            self.progress(address + i, 100.0 * i / len(bin))

        got = self.checksum(0x10040000, len(bin))
        expect = sum(ord(x) for x in bin) & 0xFFFF
    
        if got != expect:
            raise LoaderException('Checksum error')

    def interrupt(self):
        """Send an interrupt to the loader"""
        self._s.write('\003')

    def reset(self):
        """Reset the target"""
        self._s.setDTR(False)
        time.sleep(0.5)

        end = time.time() + 6

        while time.time() < end:
            self.interrupt()
            if self._s.read() == '!':
                break
        else:
            raise LoaderException('No response after reset')

        self._s.read(4096)
        self.interrupt()
        self.get()

    def progress(self, address, percent):
        """Prints out the progress.  Override for other versions."""
        print '%08x (%.1f %%)\r' % (address, percent)
        sys.stdout.flush()

def main():
    l = Loader('/dev/ttyUSB0')
    dat = open(sys.argv[1], 'rb').read()
    l.reset()
    l.load(0x10040000, dat)
    l.go(0x10040000)

if __name__ == '__main__':
    main()

